internal/controller/compaction/reconciler.go

```diff
@@ -274,6 +277,102 @@ func (r *Reconciler) delete(ctx context.Context, logger logr.Logger, etcd *druid
 	}, nil
 }
 
+func (r *Reconciler) createCompactionJobOrTriggerFullSnapshot(ctx context.Context, logger logr.Logger, etcd *druidv1alpha1.Etcd, accumulatedEtcdRevisions int64, compactionJobName string, job *batchv1.Job) (ctrl.Result, error) {
+	var err error
+	eventsThreshold := r.config.EventsThreshold
+	triggerFullsnapshotThreshold := r.config.TriggerFullSnapshotThreshold
+	if compactionSpec := etcd.Spec.Backup.SnapshotCompaction; compactionSpec != nil {
+		if compactionSpec.EventsThreshold != nil {
+			eventsThreshold = *compactionSpec.EventsThreshold
+		}
+		if compactionSpec.TriggerFullSnapshotThreshold != nil {
+			triggerFullsnapshotThreshold = *compactionSpec.TriggerFullSnapshotThreshold
+		}
+	}
+
+	// Trigger full snapshot if the delta revisions over the last full snapshot are more than the configured upper threshold
+	// or if the last job completion reason is DeadlineExceeded.
+	// This is to ensure that we avoid spinning up compaction jobs even when we know that the probability of it getting succeeded is very low due to the large number of revisions.
+	// This avoids unnecessary resource consumption and delays in the compaction process
+	if (lastJobCompletionReason != nil && *lastJobCompletionReason == batchv1.JobReasonDeadlineExceeded) || accumulatedEtcdRevisions >= triggerFullsnapshotThreshold {
```

Comment:
if the last job failed due to deadline exceeded and you have not yet crossed the `fullSnapshotEventThreshold` and now you trigger a full snapshot then you are effectively not honoring the full snapshot threshold. Job failing due to deadline exceeded is not certainly an indication that the compaction job will now never succeed unless a full snapshot is taken. This is why you introduced `fullSnapshotEventThreshold` in the first place.